File: golang/app/library/Redis.go
package library

import (
	"golang/app"
	"strconv"
	"time"

	redigo "github.com/gomodule/redigo/redis"
)

/* 缓存数据库 */
type Redis struct {
	Pool *redigo.Pool
	Conn redigo.Conn
}

func (c *Redis) Run() *Redis {
	if c.Pool == nil {
		cfg := app.Redis() //配置
		min, _ := strconv.Atoi(cfg["min"])
		max, _ := strconv.Atoi(cfg["max"])
		pool := &redigo.Pool{
			MaxIdle:     min,              //空闲数
			MaxActive:   max,              //最大数
			IdleTimeout: 10 * time.Second, //空闲超时(秒)
			Wait:        true,             //超过最大连接数: true 等待 false 报错
			Dial: func() (redigo.Conn, error) {
				c, err := redigo.Dial("tcp", cfg["host"]+":"+cfg["port"])
				if err != nil {
					panic(err)
				}
				// 密码
				if cfg["password"] != "" {
					if _, err := c.Do("AUTH", cfg["password"]); err != nil {
						c.Close()
						panic(err)
					}
				}
				// 硬盘
				if _, err := c.Do("SELECT", cfg["db"]); err != nil {
					c.Close()
					panic(err)
				}
				return c, err
			},
			TestOnBorrow: func(c redigo.Conn, t time.Time) error {
				_, err := c.Do("PING")
				panic(err)
			},
		}
		c.Pool = pool
	}
	c.Conn = c.Pool.Get()
	return c
}

/* 关闭 */
func (r Redis) Close() {
	r.Conn.Close()
	// r.Pool.Close()
}

/* 添加 */
func (r Redis) Set(name string, v interface{}) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("SET", name, v))
	return res, err
}

/* 获取 */
func (r Redis) Get(name string) ([]byte, error) {
	res, err := redigo.Bytes(r.Conn.Do("Get", name))
	return res, err
}

/* 删除 */
func (r Redis) Del(keys ...interface{}) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("DEL", keys...))
	return res, err
}

/* 是否存在 */
func (r Redis) Exist(name string) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("EXISTS", name))
	return res, err
}

/* 设置过期时间(秒) */
func (r Redis) Expire(name string, lifeTime int64) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("EXPIRE", name, lifeTime))
	return res, err
}

/* 获取过期时间(秒) */
func (r Redis) Ttl(name string) (int64, error) {
	res, err := redigo.Int64(r.Conn.Do("TTL", name))
	return res, err
}

/* 获取长度 */
func (r Redis) StrLen(name string) (int, error) {
	res, err := redigo.Int(r.Conn.Do("STRLEN", name))
	return res, err
}

/* 哈希(Hash)-添加 */
func (r Redis) HSet(name string, key string, v interface{}) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("HSET", name, key, v))
	return res, err
}
func (r Redis) HMSet(name string, obj interface{}) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("HSET", redigo.Args{}.Add(name).AddFlat(&obj)...))
	return res, err
}

/* 哈希(Hash)-获取 */
func (r Redis) HGet(name string, key string) ([]byte, error) {
	res, err := redigo.Bytes(r.Conn.Do("Get", name, key))
	return res, err
}
func (r Redis) HMget(name string, keys ...string) ([]interface{}, error) {
	args := []interface{}{name}
	for _, field := range keys {
		args = append(args, field)
	}
	res, err := redigo.Values(r.Conn.Do("HMGET", args))
	return res, err
}

/* 哈希(Hash)-删除 */
func (r Redis) Hdel(name string, key string) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("HDEL", name, key))
	return res, err
}

/* 哈希(Hash)-是否存在 */
func (r Redis) HExist(name string, key string) (bool, error) {
	res, err := redigo.Bool(r.Conn.Do("HEXISTS", name, key))
	return res, err
}

/* 哈希(Hash)-Key个数 */
func (r Redis) HLen(name string) (int, error) {
	res, err := redigo.Int(r.Conn.Do("HLEN", name))
	return res, err
}

File: php/library/Redis.php
<?php
namespace Library;

use Service\Base;
use Config\Redis as Cfg;

/* 缓存数据库 */
class Redis extends Base {

  static $RedisDB = null;       //默认池
  static $RedisDBOther = null;  //其它池

  private $db = '';             //数据库
  private $conn = null;         //连接

  /* 构造函数 */
  function __construct(string $db='') {
    $this->db = $db;
    $this->RedisConn();
  }

  /* 连接 */
  function RedisConn() {
    if($this->db=='other'){
      if(!Redis::$RedisDBOther) Redis::$RedisDBOther=$this->RedisPool(Cfg::Other());
      $this->conn = self::$RedisDBOther;
    } else {
      if(!Redis::$RedisDB) Redis::$RedisDB=$this->RedisPool(Cfg::Default());
      $this->conn = self::$RedisDB;
    }
    return $this->conn;
  }

  /* 关闭 */
  function Close() {
    if($this->conn) $this->conn->close();
  }

  /* 数据池 */
  function RedisPool(array $cfg){
    try{
      $redis = new \Redis();
      $redis->pconnect($cfg['host'], $cfg['port']); 
      if($cfg['password']) $redis->auth($cfg['password']);
      $redis->select($cfg['db']);
      return $redis;
    }catch (\Exception $e){
      self::Print('[Redis] Conn: 请检测Redis是否启动!');
      return null;
    }
  }

  /* 添加 */
  function Set(string $key, string $val) {
    if(!$this->conn) return;
    return $this->conn->set($key, $val);
  }
  /* 自增 */
  function Incr(string $key) {
    if(!$this->conn) return;
    return $this->conn->incr($key);
  }
  /* 获取 */
  function Gets(string $key) {
    if(!$this->conn) return;
    return $this->conn->get($key);
  }
  /* 删除 */
  function Del(string ...$key) {
    if(!$this->conn) return;
    return $this->conn->del($key);
  }
  /* 是否存在 */
  function Exist(string $key) {
    if(!$this->conn) return;
    return $this->conn->exists($key);
  }
  /* 设置过期时间(秒) */
  function Expire(string $key, int $ttl) {
    if(!$this->conn) return;
    return $this->conn->expire($key, $ttl);
  }
  /* 获取过期时间(秒) */
  function Ttl(string $key) {
    if(!$this->conn) return;
    return $this->conn->ttl($key);
  }
  /* 获取长度 */
  function StrLen(string $key) {
    if(!$this->conn) return;
    return $this->conn->strlen($key);
  }

  /* 哈希(Hash)-添加 */
  function HSet(string $name, string $key, $val) {
    if(!$this->conn) return;
    return $this->conn->hset($name, $key, $val);
  }
  function HMSet(string $name, array $obj) {
    if(!$this->conn) return;
    return $this->conn->hmset($name, $obj);
  }
  /* 哈希(Hash)-获取 */
  function HGet(string $name, string $key) {
    if(!$this->conn) return;
    return $this->conn->hget($name, $key);
  }
  function HMGet(string $name, string $key) {
    return $this->conn->hmget($name, $key);
  }
  /* 哈希(Hash)-删除 */
  function HDel(string $name, string ...$key) {
    if(!$this->conn) return;
    return $this->conn->hdel($name, $key);
  }
  /* 哈希(Hash)-是否存在 */
  function HExist(string $name, string $key) {
    if(!$this->conn) return;
    return $this->conn->hexists($name, $key);
  }
  /* 哈希(Hash)-Key个数 */
  function HLen(string $name) {
    if(!$this->conn) return;
    return $this->conn->hlen($name);
  }

  /* 列表(List)-写入 */
  function RPush(string $key, $val) {
    if(!$this->conn) return;
    return $this->conn->rpush($key, $val);
  }
  function LPush(string $key, $val) {
    if(!$this->conn) return;
    return $this->conn->lpush($key, $val);
  }

  /* 列表(List)-读取 */
  function LRange($key, $start, $end): array {
    if(!$this->conn) return null;
    return $this->conn->lRange($key, $start, $end);
  }
  function RPop($key) {
    if(!$this->conn) return null;
    return $this->conn->rPop($key);
  }
  function LPop($key) {
    if(!$this->conn) return null;
    return $this->conn->lPop($key);
  }
  function BRPop($key, $timeout): array {
    if(!$this->conn) return null;
    return $this->conn->brPop($key, $timeout);
  }
  function BLPop($key, $timeout): array {
    if(!$this->conn) return null;
    return $this->conn->blPop($key, $timeout);
  }

}